File: problems/11-heaps/002-top-k-frequent-elements/java/Solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

public class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> frequency = new HashMap<>();

        for (int value : nums) {
            frequency.put(value, frequency.getOrDefault(value, 0) + 1);
        }

        // Min-heap ordered by frequency.
        PriorityQueue<Integer> heap = new PriorityQueue<>(
                (a, b) -> Integer.compare(frequency.get(a), frequency.get(b))
        );

        for (int value : frequency.keySet()) {
            heap.offer(value);

            if (heap.size() > k) {
                heap.poll();
            }
        }

        int[] answer = new int[k];
        for (int i = k - 1; i >= 0; i--) {
            answer[i] = heap.poll();
        }

        return answer;
    }
}

File: problems/11-heaps/002-top-k-frequent-elements/python/solution.py
from collections import Counter
import heapq


class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        frequency = Counter(nums)

        # Keep only the k most frequent values.
        heap = []
        for value, count in frequency.items():
            heapq.heappush(heap, (count, value))

            if len(heap) > k:
                heapq.heappop(heap)

        return [value for count, value in heap]
